rct_optimizations/include/rct_optimizations/dh_chain_kinematic_calibration.h

```diff
@@ -229,15 +283,93 @@ class DualDHChainCost2D3D
+       const DHChain &target_chain,
+       const Eigen::VectorXd &camera_chain_joints,
+       const Eigen::VectorXd &target_chain_joints)
+    : DualDHChainCost (camera_chain, target_chain, camera_chain_joints, target_chain_joints)
+    , measurement_camera_to_target_(measurement)
+  {
+  }
 
-  Eigen::VectorXd camera_chain_joints_;
-  Eigen::VectorXd target_chain_joints_;
+  template<typename T>
+  bool operator()(T const *const *parameters, T *residual) const
+  {
+    // Step 1: Load the data
+    // The first parameter is a pointer to the DH parameter offsets of the camera kinematic chain
+    Eigen::Map<const Eigen::Matrix<T, Eigen::Dynamic, 4>> camera_chain_dh_offsets(parameters[0], camera_chain_.dof(), 4);
+
+    // The next parameter is a pointer to the DH parameter offsets of the target kinematic chain
+    Eigen::Map<const Eigen::Matrix<T, Eigen::Dynamic, 4>> target_chain_dh_offsets(parameters[1], target_chain_.dof(), 4);
+
+    // The next two parameters are pointers to the position and angle axis of the transform from the camera mount to the camera
+    std::size_t cm_to_c_idx = 2;
+    const Isometry3<T> camera_mount_to_camera = createTransform(parameters, cm_to_c_idx);
+
+    // The next two parameters are pointers to the position and angle axis of the transform from the target mount to the target
+    std::size_t tm_to_t_idx = cm_to_c_idx + 2;
+    const Isometry3<T> target_mount_to_target = createTransform(parameters, tm_to_t_idx);
+
+    // The next two parameters are pointers to the position and angle axis of the transform from the camera chain base to the target chain base
+    std::size_t cb_to_tb_idx = tm_to_t_idx + 2;
+    const Isometry3<T> camera_base_to_target_base = createTransform(parameters, cb_to_tb_idx);
+
+    // Step 2: Transformation math
+    // Build the transforms from the camera chain base out to the camera
+    Isometry3<T> camera_chain_fk = camera_chain_.getFK<T>(camera_chain_joints_.cast<T>(),
+                                                          camera_chain_dh_offsets);
+    Isometry3<T> camera_base_to_camera = camera_chain_fk * camera_mount_to_camera;
+
+    // Build the transforms from the camera chain base out to the target
+    Isometry3<T> target_chain_fk = target_chain_.getFK<T>(target_chain_joints_.cast<T>(),
+                                                          target_chain_dh_offsets);
+    Isometry3<T> camera_base_to_target = camera_base_to_target_base * target_chain_fk
+                                         * target_mount_to_target;
+
+    // Now that we have two transforms in the same frame, get the difference between the expected and observed pose of the target
+    Isometry3<T> camera_to_target = camera_base_to_camera.inverse() * camera_base_to_target;
+
+    Isometry3<T> camera_to_target_measured = measurement_camera_to_target_.camera_to_target.cast<T>();
+
+    Isometry3<T> tform_error = camera_to_target_measured * camera_to_target.inverse();
```

Comment:
It would probably be good to share this part of the function between both 2D-3D and this 6D cost function, since it's the same up to this point. I did something similar for the hand-eye cost function [here](https://github.com/Jmeyer1292/robot_cal_tools/blob/2ca2f0196bfb2563c84c6f5dd4c1ae40a96da685/rct_optimizations/src/rct_optimizations/extrinsic_hand_eye.cpp#L44)